protocol/0082-RFPR-on_chain_referral_program.md

```diff
@@ -0,0 +1,235 @@
+
+```psuedo
+Given:
+    benefit_tiers=[
+        {
+            "minimum_running_volume": 10000,
+            "referral_reward_factor": 0.001,
+            "referral_discount_factor": 0.001,
+        },
+        {
+            "minimum_running_volume": 20000,
+            "referral_reward_factor": 0.005,
+            "referral_discount_factor": 0.005,
+        },
+        {
+            "minimum_running_volume": 30000,
+            "referral_reward_factor": 0.010,
+            "referral_discount_factor": 0.010,
+        },
+    ]
+
+And:
+    team_running_volume=22353
+
+Then:
+    referral_reward_factor=0.005
+    referral_discount_factor=0.005
+```
+
+These benefit factors are then fixed for the duration of the next epoch.
+
+### Applying benefit factors
+
+Whenever a fee is incurred by a referee (either during continuous trading or on auction exit) the network must apply referral program benefits.
+
+The network can first calculate individual fee components following the [fees specification](./0029-FEES-fees.md#calculating-fees) and apply any usual checks.
+
+The network must then:
+
+- Calculate the rewards due to the referrer.
+    ```pseudo
+    infrastructure_fee_reward = floor(infrastructure_fee * referral_reward_factor) 
+    liquidity_fee_reward = floor(liquidity_fee * referral_reward_factor) 
+    maker_fee_reward = floor(maker_fee * referral_reward_factor) 
+    ```
+- Calculate the discounts due to the referee.
+    ```pseudo
+    infrastructure_fee_discount = floor(infrastructure_fee * referral_discount_factor)
+    liquidity_fee_discount = floor(liquidity_fee * referral_discount_factor)
+    maker_fee_discount = floor(maker_fee * referral_discount_factor)
+    ```
+- And then update the fees.
+    ```pseudo
+    infrastructure_fee = infrastructure_fee - infrastructure_fee_reward - infrastructure_fee_discount
+    liquidity_fee = liquidity_fee - liquidity_fee_reward - liquidity_fee_discount
+    maker_fee = maker_fee - maker_fee_reward - maker_fee_discount
+    ```
+
+Note the rewards and discounts are floored rather than raised to ensure the final infrastructure fee cannot be negative.
+
+The network can then carry out the normal fee transfers using the updated fee amounts followed by additional transfers from the referees general account to the referrers general account. These transfers will use the following new transfer types.
+
+- `TRANSFER_TYPE_MAKER_FEE_REWARD_PAY`
+- `TRANSFER_TYPE_MAKER_FEE_REWARD_DISTRIBUTE`
+- `TRANSFER_TYPE_LIQUIDITY_FEE_REWARD_PAY`
+- `TRANSFER_TYPE_LIQUIDITY_FEE_REWARD_DISTRIBUTE`
+- `TRANSFER_TYPE_INFRASTRUCTURE_FEE_REWARD_PAY`
+- `TRANSFER_TYPE_INFRASTRUCTURE_FEE_REWARD_DISTRIBUTE`
```

Comment:
Do we need the pay and distribute side of the types if it's done in one transfer rather than to some holding bucket?

Reply:
Yep, good point can remove the distributes thanks 👍